### DISCUSSION TITLE
base64 encode decode

### DISCUSSION THREAD

**Discussion by Sanix-Darker** [bxparks/AceCommon#32] [Ideas]
Hey there, i really like the project and am using it at the moment !
I was wondering if there are any plans on adding a base64Encode/Decode utils to it ?
I can make a PR for sure, but globally am thinking of something like this : 

```cpp
// Base64 character set
static const char base64Chars[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

// Base64 encoding function
String base64Encode(const uint8_t* data, size_t len) {
  String encoded;
  uint8_t buffer[4];

  for (size_t i = 0; i < len; i += 3) {
    buffer[0] = data[i] >> 2;
    buffer[1] = ((data[i] & 0x03) << 4) | (data[i + 1] >> 4);
    buffer[2] = ((data[i + 1] & 0x0F) << 2) | (data[i + 2] >> 6);
    buffer[3] = data[i + 2] & 0x3F;

    for (int j = 0; j < 4; j++) {
      if (i + j - len >= 0) {
        buffer[j] = 64; // Pad with '='
      }
      encoded += base64Chars[buffer[j]];
    }
  }

  return encoded;
}

// Base64 decoding function
String base64Decode(const String& encoded) {
  String decoded;
  uint8_t buffer[4];

  for (size_t i = 0; i < encoded.length(); i += 4) {
    for (int j = 0; j < 4; j++) {
      buffer[j] = encoded.charAt(i + j);
      buffer[j] = strchr(base64Chars, buffer[j]) - base64Chars;
    }

    uint8_t data[3];
    data[0] = (buffer[0] << 2) | (buffer[1] >> 4);
    data[1] = (buffer[1] << 4) | (buffer[2] >> 2);
    data[2] = (buffer[2] << 6) | buffer[3];

    for (int j = 0; j < 3; j++) {
      if (buffer[j + 1] == 64) { // Skip padding
        break;
      }
      decoded += (char)data[j];
    }
  }

  return decoded;
}
```

and to use : 
```cpp
char* originalString = "Hello, Base64!";
size_t len = strlen(originalString);
Serial.println(originalString);

// encode
String encodedString = base64Encode((const uint8_t*)originalString, len);
Serial.println(encodedString);

// decode
String decodedString = base64Decode(encodedString);
Serial.println(decodedString);
```

Let me know what you think, thanks 🙏🏽

**Comment by bxparks** [comment]
Good idea to open a ticket for discussion, before doing lots of work.

My question for you is: What is your motivation for wanting to include your code into AceCommon?
1) If you just want a base64 routine I see at least [2 existing base64 libraries](https://cse.google.com/cse?cx=008830049033802121810%3Azgsyidwem5w&q=base64) in the [Arduino Libraries](https://www.arduinolibraries.info/libraries).
2) If you need it just for your own projects, there is no need to publish it into AceCommon.
3) Is it because you can write a better implementation of base64 than the existing libraries, and you want to share that with other people?

If the rationale is (3), then you need to know that I have a pretty high bar for code quality in my libraries. So I'm going to have opinions on the efficiency of the algorithms and the coding style, which many people don't seem to want to deal with. In addition, if the purpose of the code is to be useful for other people, then writing the code is only 30% of the work. The other 70% is the unit testing, performance testing, and writing the documentation. In my experience, almost no one in the Arduino community ~~cares about those~~ wants to do those things.

Scanning through your code snippet, a few things immediately jump out at me:
1) I have a rule that my libraries should never do heap allocations. Which means, my libraries should never use the`String` class. The `String` class is terribly implemented in most Arduino platforms, with bugs continuing to be discovered year by year. So my rule, never use the Arduino String class, unless forced to do so by other 3rd party libraries.
2) Even if my library were to allow the use of `String`, you don't use the `reserve()` method on the `String` class, so your code will do multiple heap reallocations, which is highly inefficient.
3) Another potential issue in your code, you use the `strchr()` function in your decode. That makes your algorithm an `O(N*M)` algorithm, which again, makes it less efficient than it could be.
4) A small coding issue that I would want handled differently, you handle the padding of base64 using an `if` statement inside the `for` inner loop, which is bad for efficiency. The better way is to handle the padding outside the inner loop.
5) You have a `for` loop inside a `for` loop to copy the 4 base64 characters. Which means you are copying these bytes twice, which is unnecessary. I would recommend removing the extraneous inner `for` loop, so that you copy those bytes only once.

Now, if you look inside the other 2 base64 libraries that I referenced above, they suffer from a number of code quality problems, including some that I mention above. But I don't want to add another base64 implementation into AceCommon that has the same issues as the other libraries. I would want my version to be better, smaller, faster, better tested, etc.

I don't know, maybe this is a terrible way to say, no I would not accept your code as you sketched it out. Maybe you want to try again, but if you are not willing to write the unit tests, and the documentation, then I would have to say no again, because it just means more work for me. Sorry.
